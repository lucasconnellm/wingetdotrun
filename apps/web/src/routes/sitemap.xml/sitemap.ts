export type SitemapEntry = {
	url: string;
	updatedAt?: Date;
	priority?: number;
};

export const generateSitemap = (entries: SitemapEntry[]): string => {
	let sitemap = `<?xml version="1.0" encoding="UTF-8"?>\n`;
	sitemap += `<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n`;

	for (const entry of entries) {
		sitemap += `  <url>\n`;
		sitemap += `    <loc>${entry.url}</loc>\n`;
		entry.updatedAt && (sitemap += `    <lastmod>${entry.updatedAt.toISOString()}</lastmod>\n`);
		sitemap += `    <priority>${entry.priority ?? 0.5}</priority>\n`;
		sitemap += `  </url>\n`;
	}

	sitemap += `</urlset>`;
	return sitemap;
};
