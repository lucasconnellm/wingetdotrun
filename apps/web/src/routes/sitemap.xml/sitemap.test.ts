import { describe, expect, it } from "vitest";
import { generateSitemap } from "./sitemap";

describe("generateSitemap", () => {
	it("should generate a sitemap", () => {
		const entries = [
			{
				url: "https://example.com",
				updatedAt: new Date("2023-01-01"),
				priority: 1,
			},
		];
		const sitemap = generateSitemap(entries);
		expect(sitemap).toMatchSnapshot();
	});

	it("should generate a sitemap with multiple entries", () => {
		const entries = [
			{
				url: "https://example.com",
				updatedAt: new Date("2023-01-01"),
				priority: 1,
			},
			{
				url: "https://example.com/about",
				updatedAt: new Date("2023-01-02"),
			},
			{
				url: "https://example.com/careers",
				priority: 0.25,
			},
		];
		const sitemap = generateSitemap(entries);
		expect(sitemap).toMatchSnapshot();
	});
});
