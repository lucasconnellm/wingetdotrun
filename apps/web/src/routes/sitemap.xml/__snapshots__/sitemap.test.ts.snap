// Vitest Snapshot v1

exports[`generateSitemap > should generate a sitemap 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\">
  <url>
    <loc>https://example.com</loc>
    <lastmod>2023-01-01T00:00:00.000Z</lastmod>
    <priority>1</priority>
  </url>
</urlset>"
`;

exports[`generateSitemap > should generate a sitemap with multiple entries 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\">
  <url>
    <loc>https://example.com</loc>
    <lastmod>2023-01-01T00:00:00.000Z</lastmod>
    <priority>1</priority>
  </url>
  <url>
    <loc>https://example.com/about</loc>
    <lastmod>2023-01-02T00:00:00.000Z</lastmod>
    <priority>0.5</priority>
  </url>
  <url>
    <loc>https://example.com/careers</loc>
    <priority>0.25</priority>
  </url>
</urlset>"
`;
